src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/FallbackRazorProjectHost.cs

```diff
@@ -0,0 +1,143 @@
+                suppressVersionOnlyUpdates: true,
+                ruleNames: new string[] { ResolvedCompilationReference.SchemaName });
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            await base.DisposeCoreAsync(initialized).ConfigureAwait(false);
+
+            if (initialized)
+            {
+                _subscription.Dispose();
+            }
+        }
+
+        // Internal for testing
+        internal async Task OnProjectChanged(IProjectVersionedValue<IProjectSubscriptionUpdate> update)
+        {
+            await ExecuteWithLock(async () =>
+            {
+                if (IsDisposing || IsDisposed)
+                {
+                    return;
+                }
+
+                string mvcReferenceFullPath = null;
+                var references = update.Value.CurrentState["ResolvedCompilationReference"].Items;
+                foreach (var reference in references)
+                {
+                    if (reference.Key.EndsWith(MvcAssemblyFileName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        mvcReferenceFullPath = reference.Key;
+                        break;
+                    }
+                }
+
+                if (mvcReferenceFullPath == null)
+                {
+                    // Ok we can't find an MVC version. Let's assume this project isn't using Razor then.
+                    await UpdateProjectUnsafeAsync(null).ConfigureAwait(false);
+                    return;
+                }
+
+                var version = GetAssemblyVersion(mvcReferenceFullPath);
+                if (version == null)
+                {
+                    // Ok we can't find an MVC version. Let's assume this project isn't using Razor then.
+                    await UpdateProjectUnsafeAsync(null).ConfigureAwait(false);
+                    return;
+                }
+
+                var languageVersion = version.Major + "." + version.Minor;
+
+                var hostProject = new HostProject(CommonServices.UnconfiguredProject.FullPath, languageVersion);
+                await UpdateProjectUnsafeAsync(hostProject).ConfigureAwait(false);
+            });
+        }
+
+        // virtual for overriding in tests
+        protected virtual Version GetAssemblyVersion(string filePath)
+        {
+            return ReadAssemblyVersion(filePath);
+        }
+
+        private static Version ReadAssemblyVersion(string filePath)
+        {
+            try
+            {
+                using (var stream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete))
+                using (var reader = new PEReader(stream))
+                {
+                    var metadataReader = reader.GetMetadataReader();
+
+                    var assemblyDefinition = metadataReader.GetAssemblyDefinition();
+                    return assemblyDefinition.Version;
+                }
+            }
```

Comment:
We're going to disk now to read the assembly version. Using the service to do it still has an ordering problem. Since we're listening to evaluation changes from MSBuild we're still running to early to ask the assembly service for its version.